### ISSUE TITLE
More faithful implementation of U-net?

### ISSUE THREAD

**Issue by L0SG** [meetshah1995/pytorch-semseg#21]
Hi, thank you for a great work @meetshah1995 ! I discovered this repo while searching for U-net in pytorch.

While reviewing the code, I've found that there are some subtle differences than the reference U-net model in the paper (as mentioned in #1).

 1. making conv and maxpool in downsampling separate variables would enable concatenation of upsampling layer and the tensor before maxpool.

        conv1 = self.conv1(inputs)
        maxpool1 = self.maxpool1(conv1)
        conv2 = self.conv2(maxpool1)
        maxpool2 = self.maxpool2(conv2)
        conv3 = self.conv3(maxpool2)
        maxpool3 = self.maxpool3(conv3)
        conv4 = self.conv4(maxpool3)
        maxpool4 = self.maxpool4(conv4)
        center = self.center(maxpool4)

        # upsampling with skip connection from downsampling layers
        up4 = self.up_concat4(conv4, center)
        up3 = self.up_concat3(conv3, up4)
        up2 = self.up_concat2(conv2, up3)
        up1 = self.up_concat1(conv1, up2)

2. I think the paper doesn't use padding in conv (unetConv2 class in the code). changing the padding parameter from 1 to 0 would fix the concatenation dimension mismatch error too.
```
        if is_batchnorm:
            self.conv1 = nn.Sequential(nn.Conv2d(in_size, out_size, 3, 1, 0),
                                       nn.BatchNorm2d(out_size),
                                       nn.ReLU(),)
            self.conv2 = nn.Sequential(nn.Conv2d(out_size, out_size, 3, 1, 0),
                                       nn.BatchNorm2d(out_size),
                                       nn.ReLU(),)
        else:
            self.conv1 = nn.Sequential(nn.Conv2d(in_size, out_size, 3, 1, 0),
                                       nn.ReLU(),)
            self.conv2 = nn.Sequential(nn.Conv2d(out_size, out_size, 3, 1, 0),
                                       nn.ReLU(),)
```

3. Assuming that the input image is square, cropping (i.e. F.pad with negative values for crop) would be
`padding = 2 * [offset // 2, offset // 2]`  ? 

4. the stride size in unetUp class is currently 1, but I think it should be 2 (which means upsample by factor 2)
```
if is_deconv:
            # fix: add stride parameter of 2 (which is upsample by 2)
            self.up = nn.ConvTranspose2d(in_size, out_size, kernel_size=2, stride=2)
        else:
            self.up = nn.UpsamplingBilinear2d(scale_factor=2)
```

Currently testing the modified model in my [repo](https://github.com/L0SG/Liver_segmentation/blob/master/lib/unet.py) and seeing the expected behavior with CT dataset.

What's your thoughts? would PR if you think it's right. Thanks

**Comment by asanakoy** [comment]
Hey, Completely agree with everything you have written. 
The code for Unet is not working  as it is in the repository now on `pytorch 0.2.0_1`.
It has to be updated

**Comment by meetshah1995** [comment]
@L0SG - Yes you're right and apology for late reply. It would be awesome if you could send in a PR with the updated and correct U-Net implementation.  

@asanakoy - I will at some point update all the code in the repo to the latest PyTorch version. Feel free to send in a PR. 

Keeping the issue open, will close it after the PR merge.

**Comment by meetshah1995** [comment]
Fixed by #35
